Merge new weather files in to develop branch

jermainegug opened the pull request.

jermainegug committed:
Add metdata weather & local config file

jermainegug committed:
Adjust comments and documentation

jermainegug committed:
Start to form base weather class

* minor comment change in weather_ext.py
* delete the check past 15 minute for rain in weather_ext.py
* create and make common function in weather_abstract.py

jermainegug committed:
Write some code in abstract weather

* edit the weather files to work with abstact weather
* add stuff to config file to make more generic

jermainegug committed:
Import ABC

jermainegug committed:
Multiple products in config

* import `get_mongodb` in weather files
* put libraries back into `weather_ext.py`

jermainegug committed:
Edit weather files

jermainegug committed:
Edit capture method

jermainegug committed:
Allow for data to be included in safety decision

* for the safety condition methods the current_values parameter was included so that the metdata does not have to be fetched each time capture calls on them

NOTE: If the base class doesn't work use the code at this stage to use with the dome.

jermainegug committed:
Merge branch 'improve-weather-code' into emergency-weather-monitor

jermainegug committed:
post-merge cleanup

* better base methods
* more generic
* should work

jermainegug committed:
Check to make sure num col = num units specified

jermainegug committed:
Merge branch 'improve-weather-code' into origin/improve-weather-code

AnthonyHorton commented on config.yaml:
```diff
@@ -13,7 +13,63 @@ environment:
     station: mongo
+    web_service:
+      name: 'aat-metdata'
+      link: 'http://site.aao.gov.au/met/metdata.dat'
```
As per Steve Lee's email Re: AAT met data table format on 25/11/17 should be using `
ftp://site-ftp.aao.gov.au/pub/local/metdata/metadata1.dat` instead. The latter is updated every minute instead of every 10 minutes, and has simpler Boltwood rain & wet flats (just 0 or 1, instead of number of minutes with positive readings in last 10).

jermainegug replied:
Link is changed.
I am assuming that the order of the elements in the file are the same as the last one.

AnthonyHorton replied:
Format is the same, but the rain & wet flags are just -1, 0 or 1 (no data, not wet, wet) instead of integers 0 to 10 (number of wet readings in the last 10 minutes).

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+        self.table_data = self.fetch_met_data()
+        col_names = self.web_data.get('column_names')
+        for i in range(0, len('col_names')):
+            data[col_names[i]] = self.table_data[col_names[i]][0]
```
No need for a loop counter here. There rarely is in Python, and when you really need one you should use the built in `enumerate()` function instead.  This is neater:

    for name in col_names:
        data[name] = self.table_data[name][0]

jermainegug replied:
Has been changed.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+        try:
+            cache_age = Time.now() - self._met_data['Time (UTC)'][0] * 86400
+        except AttributeError:
+            cache_age = self.web_data.get('cache_age', 1.382e10) * u.year
```
No need to go checking the config file for an alternative default value for `cache_age`. The only purpose of this is to ensure that if there is no cache yet (i.e. first call to `fetch_met_data()`) then the following `if cache_age > self.max_age:` line will return `True` and download met data. That's why I made the default value of `cache_age` equal the age of the Universe...

There are more elegant ways to do this without faking a cache age, but this was my quick solution.

jermainegug replied:
Yep, I have gotten rid of the check for a value for `cache_age`. Don't know any other ways to make it better.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+            # Change string format to ISO
+            t['Time (UTC)'].format = 'iso'
+            # Convert from AAT standard time to UTC
+            t['Time (UTC)'] = t['Time (UTC)'] - 10 * u.hour
```
A note for the future, this time munging stuff is very, very specific to the odd time stamp format of the AAT met data. Ideally these quirks would be hidden away in an even lower level subclass, but this is OK for now.

jermainegug replied:
Once the code is running and working well, we can do that. Would it just be making a subclass that contains the code to change the time format and then calling the class into the current one?
Just trying to understand how it would work. :)

AnthonyHorton replied:
Yes, something like that. A class that parses sensibly formatted tables of met data in a fairly general, adaptable way, and a subclass of that that adds the extra processing steps specific to the AAT (pruning surplus newlines & quotation marks, manual timezone conversion from AAT standard time to UTC).

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+
+            # Set units for items that have them
+            for i in range(1, len('col_names')):
+                t[col_names[i]].unit = col_units[i]
```
As above, manual loop counters like this are almost never the best solution in Python. You could use the `enumerate()` function:

    for i, name in enumerate(col_names):
        t[name].unit = col_units[i]

Or, even better, the `zip()` function:

    for name, unit in zip(col_names, col_units):
        t[name].unit = unit

jermainegug replied:
Has been changed.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+        entries = self.weather_entries
+
+        sky_diff = weather_entries[self.web_data.get('sky_ambient')]
+        sky_diff_u = weather_entries[self.web_data.get('sky_ambient_uncertainty')]
```
This and line above should be `entries`, not `weather_entries`, I think?

TBH using `self.weather_entries` directly for these two lines might be better.

AnthonyHorton commented on config.yaml:
```diff
@@ -13,7 +13,63 @@ environment:
+      maximum_wind_gust: 'Maximum wind gust'
+      rain_sensor: 'Rain sensor'
+      rain_flag: 'Boltwood rain flag'
+      wet_flag: 'Boltwood wet flag'
```
I'm not sure what the purpose of these config file entries are. You're basically creating an alias for some of the column names, but I don't see why that's useful.

What might be useful would be a list of the columns that tell you it's raining/wet (e.g. Rain sensor, Boltwood rain flag, Boltwood wet flag) to enable the code to just do a big logical OR on the values of those columns, but these 1 to 1 aliases don't help with that.

jermainegug replied:
I think I was trying avoid using the column names straight up for some parts of the code but I think it over complicates what I was trying to do. I have gotten rid of them and fixed up the corresponding code.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,197 @@
+        self.config = load_config()
+
+        # Read configuration
+        self.sensor_data = self.config['weather']['aag_cloud']
```
This doesn't belong in the abstract base class, it's specific to the AAG CloudWatcher.

jermainegug replied:
Has been moved.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,197 @@
+        self.sensor_data = self.config['weather']['aag_cloud']
+        self.safety_delay = self.sensor_data.get('safety_delay', 15.)
+
+        self.web_data = self.config['weather']['web_service']
```
As above, this is specific to web met data, it doesn't belong in the abstract base class.

jermainegug replied:
Has been moved.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,197 @@
+        safety_delay = self.safety_delay
+
+        threshold_cloudy = self.sensor_data.get('threshold_cloudy', -22.5)
+        threshold_very_cloudy = self.sensor_data.get('threshold_very_cloudy', -15.)
```
Threshold values should be stored in a generic class member variable (e.g. `self.safety_thresholds`), the values of which are set by the derived classes according to their own entries in the config file(s). Here you're using config entries specifically for the AAG CloudWatcher in the abstract base class, which should be generic.

jermainegug replied:
I have made a new section in the config file that holds all the threshold values.

AnthonyHorton replied:
That's not really what I meant. We should talk through this.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,197 @@
+        threshold_cloudy = self.sensor_data.get('threshold_cloudy', -22.5)
+        threshold_very_cloudy = self.sensor_data.get('threshold_very_cloudy', -15.)
+
+        if len(sky_diff) == 0:
```
`sky_diff` would be undefined here. You need to either pass `sky_diff` as an argument to this method, or make it a class member variable, e.g. `self._sky_diff`. Local variables from methods of derived classes are not automatically available in the corresponding method of the parent class.

jermainegug replied:
Is this okay? `def _get_cloud_safety(self, sky_diff, last_cloud): ...`

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,197 @@
+        threshold_gusty = self.sensor_data.get('threshold_gusty', 40)
+        threshold_very_gusty = self.sensor_data.get('threshold_very_gusty', 50)
+
+        if len(wind_gust) == 0:
```
As with cloud safety, this method won't have access to `wind_gust`, `wind_speed`, etc., unless you pass them as arguments or make them member variables.

jermainegug replied:
Is this okay? `def _get_wind_safety(self, wind_speed, wind_gust): ...`

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,197 @@
+        safety_delay = self.safety_delay
+        entries = self.weather_entries
+        # since the rain data is different in the two systems
+        # can not make a method that satisfies both
```
It is possible. They both have one or more sensors that say either 'raining/wet' or 'not raining/wet'.  The method just needs to know which columns of the data mean wet, then logical OR them.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,197 @@
+
+    def capture(self):
+        # Make Safety Decision
+        self.safe_dict = self.make_safety_decision(data)
```
`data` will be undefined here. It either needs to be passed as an argument to this method, or to be a class member variable.

jermainegug replied:
Like this? `def capture(self, data) ...`

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+        for i in range(0, len('col_names')):
+            data[col_names[i]] = self.table_data[col_names[i]][0]
+
+        super.capture()
```
`data` is a local variable, only available inside this method. This call to the `capture()` method of the parent, abstract base class won't be able to access `data` unless you pass it as an argument or make it into a member variable instead.

Do a bit of background reading on variable scope in Python.

jermainegug replied:
I have added data as an argument, like so `super.capture(data)`.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+                'Sky': cloud[0],
+                'Wind': wind[0],
+                'Gust': gust[0],
+                'Rain': rain[0]}
```
You call `super.make_safety_decision()` but don't store its return values, then attempt to reference those return values in your `return` statement, which will fail.

Need either:

    safety_dict = super.make_safety_decision(current_values)
    
    return safety_dict

Or simply:

    return super.make_safety_decision(current_values)

Really this whole method might be best left to the abstract base class. The only thing you're adding here is an additional log entry, which could also be done in the base class.

jermainegug replied:
I have got the log entry in the base class but it doesn't specify what system it using.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+        max_sky_diff = sky_diff + sky_diff_u
+        last_cloud = sky_diff
+
+        super._get_cloud_safety()
```
As with comments elsewhere, this call to `super._get_cloud_safety()` isn't going to get the values of `last_cloud`, `max_sky_diff`, etc.  Need to be either passed values as arguments, or use member variables.

jermainegug replied:
Like this? `super.__get_cloud_safety(max_sky_diff, last, cloud)`

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+
+        super._get_cloud_safety()
+
+        return cloud_condition, sky_safe
```
As with `make_safety_decision` you're discarding the return values from the `super.<method_name>()` call then trying to reference them in the `return` statement.  Just `return super.<method_name()>`

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+        wind_speed = entries[self.web_data.get('average_wind_speed')]
+        wind_gust = entries[self.web_data.get('maximum_wind_gust')]
+
+        super._get_wind_safety()
```
As elsewhere, you need make the wind values available to `super._get_wind_safety()` somehow (arguments, member variables).

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+
+        super._get_wind_safety()
+
+        return (wind_condition, wind_safe), (gust_condition, gust_safe)
```
Again, need to capture these variables from the `super._get_wind_safety()` call above if you want to `return` them.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,188 @@
+            rain_condition = 'Unknown'
+        else:
+            # Check current values
+            if rain_sensor > threshold_rain and rain_flag > threshold_rain:
```
Pretty sure we want that to be `or` instead of `and`.  If either of the two rain sensors in the AAT weather station say it's raining, we should assume it's raining.

jermainegug replied:
Has been changed.

AnthonyHorton reviewed (commented):
Looks good, but it's not quite ready to merge yet.  The separation between abstract base class and concrete derived classes is not as clear cut as it should be, and there are a number of errors relating to variable scope. See specific comments for details.

We should have another chat about the code structure, and we really need to set you up so you can test the code properly. It's basically impossible to write any non-trivial piece of code if you can't run it!

jermainegug committed:
pre-pre-merge cleanup (weather-monitor to develop)

jermainegug committed:
Fix rain safety classes

* majority of the code is in the base class
* code in `weather.py` truns the AAG sensor's rain values into binary values to work with the base class

jermainegug committed:
Add arguments to rain safety base class

jermainegug committed:
Edit thresholds in config file and make necessary changes

* rain safety check was edited to make all entries ~binary for rain condition (needs to be changed anyway)
* `simple_weather_capture.py` incldes code to be able to test on windows laptop (needs to be changed)

jermainegug committed:
Replace previous rain safety method with a draft

* adjust comments in `simple_weather_capture.py`

jermainegug committed:
Add rain safety code to get condition and safety 

* adjusted config file to allow for infinity values to work cut out the clear/calm conditions which were not used in the code.
* fix weather and weather_web_data files, correct super().'s, add rain safety checks.
* fix weather abstract file, correct methods
*import weather_web_data to the simple_weather_capture file to run both the files for aag and aat data.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,214 @@
+            wind = ['N/A']
+            gust = ['N/A']
+
+        rain = self._get_rain_safety(current_values)
```
The abstract base class doesn't (and shouldn't) have a `_get_rain_safety()` method so this will fail with an `AttributeError` in the base class or in any derived class that doesn't implement `_get_rain_safety()`.

If you think that `_get_rain_safety()` should be implemented in all derived classes (i.e. all weather stations will always have a rain sensor of some kind) then the correct way to handle this is to put a `_get_rain_safety()` method in the abstract base class that contains only `raise NotImplementedError`.  That makes it clear to someone writing a derived class that they need to write a `_get_rain_safety()` method, and ensures that there is a reasonably informative error message if they don't.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,214 @@
+
+        return data
+
+    def make_safety_decision(self, current_values):
```
This implementation of `make_safety_decision()` makes a lot of assumptions about the types of sensors that the weather station has (i.e. cloud sensors, wind sensors that give average and gust values, and rain sensors), and that's not appropriate for an abstract base class.

All this method should really be doing is calling `_get_status()` to apply the thresholds to the current weather data, then passing the resulting dictionary of statuses to a list of other methods (stored in a `self._safety_methods` member variable, say), ANDing the results to decide whether the overall safety status is 'safe' or 'not safe', and returning the results in a dictionary structure like one you're using.  That's entirely generic and independent of weather station details.

It's then up to the derived classes to specify which methods `make_safety_decision()` should call, and implement those methods using logic specific to their own set of sensors.

AnthonyHorton replied:
To facilitate this you should change the individual safety methods so that they all have the same interface.  They should all just take the status dictionary produced by `_get_status()`, extract the statuses that are relevant to their category (e.g. rain, or wind), apply whatever device specific logic is necessary (e.g. combining multiple rain sensors) and return a safe/not safe decision for that category.

jermainegug replied:
Have included `statuses = self._get_status()` and then have passed `statuses` as an argument in the `_get_..._safety()` methods.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,214 @@
+
+        return cloud_condition, sky_safe
+
+    def _get_wind_safety(self, wind_speed, wind_gust):
```
For the sake of modularity and consistency of interface with the other safety decision methods it may be better to split this into separate average wind speed and wind gust methods.

AnthonyHorton replied:
Also the wind safety methods should work with the statuses returned from `_get_status()`, not directly with the met data.

jermainegug replied:
Modified to use the results from _get_status() and I have split the wind and gust into different methods in the base class.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,214 @@
+                'Gust': gust[0],
+                'Rain': rain[0]}
+
+    def _get_cloud_safety(self, sky_diff, last_cloud):
```
Cloud safety should work with the statuses returned from `_get_status()`, not directly with the met data values. That's the point of the `_get_status()` method, it applies all the thresholds all at once in a device independent way.

jermainegug replied:
Modified to use the results from `_get_status()`.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,214 @@
+
+        return (wind_condition, wind_safe), (gust_condition, gust_safe)
+
+    def _get_status(self):
```
You need to actually use this method!

AnthonyHorton replied:
For consistency with other methods should probably take `current_values` as an argument, instead of poking around in `weather_entries` itself.

jermainegug replied:
Done.

AnthonyHorton commented on config.yaml:
```diff
@@ -33,9 +95,24 @@ weather:
+          wind_speed:
+            windy: 50., 75.
+            very_windy: 75., .inf
+          wind_gust:
```
You need a 'not gusty' (or equivalent) category e.g. `0, 50`, otherwise `_get_status()` is going to mark non-gusty conditions with the 'invalid' default status.

jermainegug replied:
Included the 'calm' category (equivalent to 'not gusty').

AnthonyHorton commented on config.yaml:
```diff
@@ -33,9 +95,24 @@ weather:
             impulse_cycle: 600 ## seconds
+        thresholds:
+          sky-ambient:
+            cloudy: -25., -15.
```
You need a 'clear' category e.g. `-.inf, -25`, otherwise `_get_status()` is going to mark clear conditions with the 'invalid' default status.

jermainegug replied:
Inlcuded the 'clear' category.

AnthonyHorton replied:
You have for the AAT met data, but 'clear', 'calm', etc. are still missing for the local AAG section.

AnthonyHorton commented on peas/weather.py:
```diff
@@ -784,175 +740,55 @@ def calculate_and_set_PWM(self):
-                'Gust': gust[0],
-                'Rain': rain[0]}
-
     def _get_cloud_safety(self, current_values):
```
Modify to use results from `_get_status()`

jermainegug replied:
Modified to use the results from `_get_status()`.

AnthonyHorton commented on peas/weather.py:
```diff
@@ -784,175 +740,55 @@ def calculate_and_set_PWM(self):
-        return cloud_condition, sky_safe
+        return super()._get_cloud_safety(max_sky_diff, last_cloud)
 
     def _get_wind_safety(self, current_values):
```
Modify to use results from `_get_status()`, consider splitting into separate average wind speed and wind gust methods.

jermainegug replied:
Modified to use the results from `_get_status()` and I have split the wind and gust into different methods in the base class.

jermainegug replied:
With the two methods I have to somehow use them to get the values for wind and gust into weather entries so that they can be used in the base methods. Any pointers?

AnthonyHorton commented on peas/weather.py:
```diff
@@ -784,175 +740,55 @@ def calculate_and_set_PWM(self):
-        return (wind_condition, wind_safe), (gust_condition, gust_safe)
+        return super()._get_wind_safety(wind_speed, wind_gust)
 
     def _get_rain_safety(self, current_values):
```
Modify to use results from `_get_status()`.

jermainegug replied:
Modified to use the results from `_get_status()`.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,170 @@
+        super().__init__(use_mongo=use_mongo)
+
+        # Read configuration
+        self.web_data = self.config['weather']['web_service']
```
I think `self.web_data` is a misleading name for this.  It's not data, for one thing, it's configuration info, specifically the configuration info for the met data web service.  `self.web_config`?

jermainegug replied:
Yep, I agree. Does seem kind of misleading, I have changed it to `self.web_config`. On this note, is the file name `weather_web_data.py` OK? I feel like I can name it something better but I can't think of it!

AnthonyHorton replied:
A pendant might point out that since the best source of AAT met data is an ftp site and not an http site then it is not in fact 'web' data... (though it is internet data). I wouldn't worry about the file name at this point, though.

jermainegug replied:
I have just changed it to `internet_weather.py`, just to avoid confusion.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,170 @@
+
+        return self._met_data
+
+    def _get_cloud_safety(self, current_values):
```
Modify to work with the statuses returned from `_get_status()`, not directly with the met data.

jermainegug replied:
Modified to use the results from `_get_status()`.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,170 @@
+
+        return super()._get_cloud_safety(max_sky_diff, last_cloud)
+
+    def _get_wind_safety(self, current_values):
```
Modify to work with the statuses returned from `_get_status()`, not directly with the met data.

jermainegug replied:
Modified to use the results from `_get_status()`.

AnthonyHorton commented on peas/weather_web_data.py:
```diff
@@ -0,0 +1,170 @@
+
+        return super()._get_wind_safety(wind_speed, wind_gust)
+
+    def _get_rain_safety(self, current_values):
```
Modify to work with the statuses returned from `_get_status()`, not directly with the met data.

jermainegug replied:
Modified to use the results from `_get_status()`.

AnthonyHorton reviewed (commented):
Still some refactoring needed to complete the split into abstract base class and concrete derived classes.

AnthonyHorton commented on peas/weather_abstract.py:
```diff
@@ -0,0 +1,214 @@
+        data['Sky condition'] = self.safe_dict['Sky']
+        data['Wind condition'] = self.safe_dict['Wind']
+        data['Gust condition'] = self.safe_dict['Gust']
+        data['Rain condition'] = self.safe_dict['Rain']
```
These 5 lines are just duplicating the contents of the dictionary returned by `make_safety_decision()` with the addition of the word 'condition'.  Just put 'condition' into the names in `self._safety_methods`  in `__init__()` and remove all this.

AnthonyHorton replied:
Also, reuse of the variable `data`, 1st as the current weather data values passes as an argument to this method, then as dictionary of safety status info returned from `make_safety_decision()`.  That's just confusing, use a different variable name for the second case, e.g. `safety_status`.

AnthonyHorton replied:
And to continue my nitpicking, `data` is a terrible variable name. Far too vague. Variable (& function, & class) names should describe the meaning (not implementation) of the thing in question.

jermainegug replied:
Yep, I added condition into the names.
Mmm, I didn't catch on to that before... Makes it much more confusing! I have fixed it up, looks like this:
```
 def capture(self, current_values):
        # Make Safety Decision and store current weather
        current_weather = self.make_safety_decision(current_values)
        self.weather_entries.append(current_weather)

        if send_message:
            self.send_message({'data': current_weather}, channel='weather')

        if use_mongo:
            self.db.insert_current('weather', current_weather)

        return current_weather
```

AnthonyHorton reviewed (commented):
Further comments.

jermainegug committed:
Add 'clear', 'calm', etc. to thresholds in config

jermainegug committed:
Make changes to code in `weather_abstract.py`

* edit capture method so that it is more understandable
* edit make_safety_decision method so that it is more concise and generic
* _get_safety methods changed to be more concise and include the invalid along with an unknown condition
* edit _get_status method back to invalid from unknown

jermainegug committed:
Fix typo's and adjust some methods

jermainegug committed:
Change file name and minor edits

* delete _get_cloud_safety and _get_wind_safety methods - they become obsolete as the methods in base class covered all aspects of derived methods

jermainegug committed:
Edit config and make changes to wind, gust and cloud methods

* not sure if the way I have implemented self.weather_entries['...'] = value, is correct. Doesn't seem right but not sure what else to do.

jermainegug committed:
Fix typo and file name change

* updated latptop: path was changed

jermainegug committed:
Correcting code in derived weather classes

jermainegug committed:
Added base method for rain safety

Note that this method is not implemented due to the AAG and AAT having different thresholds for their rain data.

jermainegug committed:
Change confusing data variable to current_values for capture method

This is for both dervied methods

jermainegug committed:
Fix typo's

jermainegug committed:
Fixing up code

jermainegug commented:
If someone could just have a read over `abstract_weather` that would be great. Having an issue with `_get_status`, I get a `KeyError` but it uses `weather_entries` before anything is assigned to it but then we need `_get_status` to put entries in `weather_entries. So I am lost, I can't quite see how it tackle this issue.
Thanks.

jermainegug committed:
Fixed errors that came up when running

* When I run the code it doesn't give me anything back - It glitches and I have to shut down the prompt window...

jermainegug commented:
Have fixed all the errors that came up but now when I run it, it doesn't finish and I have to close the prompt window. Don't know what I have done or how to fix it. I have looked over all the files I have changed numerous times but I am not getting anything.
Worried that I might not have it up and running by Thursday...

AnthonyHorton commented:
I'm not sure you're expecting it to finish, @jermainegug. Shouldn't it just run continuously, logging weather data to the file/mongoDB?

Edit: Yes. Look at https://github.com/AstroHuntsman/PEAS/blob/emergency-weather-monitor/scripts/simple_weather_capture.py, from line 158, `while True:`  This is an infinite loop (unless you specify `--loop False` when you run the script) so you should expect to run until you force it to stop (would be Ctrl-C on Linux/OS X, not sure of the Windows equivalent).  

You need to check the contents of the output file, mongoDB and, especially, the log file to work out whether it's running as intended.

jermainegug commented:
Yep, I have the date, safe, and the weather conditions that are being logged into the file.

AnthonyHorton commented:
That sounds promising! As well as the weather status log file you'll also want to check the general purpose log files that your class is creating and writing to (with all those `self.logger.info()`, `self.logger.debug()`, etc. commands).  Looks like you're setting the filename for that to be the same as the `name` field from the config.  For the purposes to see more of what's going on you can temporarily set the logging level to `logging.DEBUG` instead of `logging.INFO`, as it is now.

jermainegug commented:
Where will I see the logs? None of them are in the prompt window or the csv file.

AnthonyHorton commented:
They're in log _files_, the name of which are set by your code in the `__init__()` method of the weather classes (the `self.logger = logging.getLogger(self.web_config.get('name'))` lines).  Look for files like `<name>.log` where `name` is whatever is next to `name:` in the config file.

Read the docs: https://docs.python.org/3/library/logging.html#module-level-functions

jermainegug commented:
It creates a folder, `logs` but there are no files in them.

AnthonyHorton commented:
Probably because all your logging commands in the internet weather class are debug level, but you've set it to only log info & above level. Temporarily chance the logging level to debug, as I suggested above, and you should see log files appear.

jermainegug commented:
I'm sure I did that, I have `self.logger.setLevel(logging.DEBUG)` instead of `self.logger.setLevel(logging.INFO)`.

AnthonyHorton commented:
OK, looks like you might be missing some logging configuration steps required to get the logs to write to file. See https://docs.python.org/3.6/howto/logging.html

This might be a result of @wtgee's clean up of PEAS to prepare it for merging with POCS.  In POCS the logger set up is handled by a module wide base class, `PanBase`.  Possibly not worth expending much effort on the logger configuration as when we do merge your code with POCS it will just use the POCS logger anyway.

AnthonyHorton commented:
Two thing that we do want, though, are docstrings and unit tests.  The latter will take a bit of explaining how to do, but the former you can get started on now.

`docstrings` are specially formatted multi-line comments (i.e. surrounded by triple backticks\backquotes instead of single lines with hashes) that immediately follow the `def` line of functions & methods.  These are used for inline documentation, and if you follow a standard format they can be automatically converted into documentation webpages/PDF, brought up as inline help in Jupyter notebooks, Atom, etc., etc.

There are a few standards in common use, PANOPTES (and hence POCS) follows the Google standard: https://google.github.io/styleguide/pyguide.html?showone=Comments#Comments  Take a look at some POCS code for examples of the format to use.

Every class, method or function should have at least a one line summary docstring, and every public one should have a full docstring including at least a one line summary, a description of the arguments (if any) and a description of the return values (if any).

jermainegug commented:
I am lost with the debugging stuff, been reading through it but I am not sure how to use it.
I read a little bit about `docstrings` the other week and I have some idea of what to do but I will read through that one tomorrow to get a better idea.

AnthonyHorton commented:
Unit tests and debugging are separate things, though the former can be very helpful when you're trying to do the latter.  We'll get to them when we have more time. Writing some docstrings is a good exercise for now.

jermainegug committed:
Big clean up

* added comments - may need to be reviewed
* fixed names
* fixed up simple_weather_capture.py

jermainegug commented:
I have written some comments, but don't think they are exactly what we are looking for. I have gone through all the previous changes of the file and I can't seem to find any changes to the `logger.DEBUG`.

jermainegug committed:
Typo fixes and adjust simple weather capture

jermainegug committed:
Allow for simple weather capture to write all weather data to file

jermainegug committed:
Fixes

jermainegug committed:
Edit so that it actually works

jermainegug committed:
Working files!

jermainegug committed:
Make names more readable

*add back write header (not currently used)
*change file name for aat metdata class

jermainegug committed:
Changing from web_service to aat_metdata

If we want to eventually add more weather data, perhaps from the 2.3 m and the sky-mapper.

jermainegug committed:
Small edit and change

jermainegug committed:
Now we know what the 120. is

jermainegug committed:
Fix typo in config

jermainegug committed:
Fixes

jermainegug committed:
Name and serial port change

jermainegug committed:
Edits and some adjustment

jermainegug committed:
Change to test AAG temporarily
